Correct port-forward data copying logic

ncdc opened the pull request:
Correct port-forward data copying logic so that the server closes its
half of the data stream when socat exits, and the client closes its half
of the data stream when it finishes writing.

Modify the client to wait for both copies (client->server,
server->client) to finish before it unblocks.

Fixes #8766
Fixes #13673

@smarterclayton @derekwaynecarr @pmorie

ncdc commented:
I'm looking for a write to write an e2e or some other useful test for this... would appreciate any suggestions. We have to make sure that port-forward can handle either side finishing its write while the other side is still in flight.

smarterclayton commented:
An e2e test that started a container, then used SSH to do various terminations (sigkill the container) while holding the client open in the test would probably be the best.  You don't need to use a real client, you could do that from the go code of the e2e test.

ncdc commented:
This PR isn't quite right yet. If instead of running `echo stat | nc ...` to Zookeeper, you run `telnet` and manually type `stat` and hit enter, the telnet connection stays open. If you then try to type another command, the connection is immediately closed because the `socat` process that's forwarding from the node to the pod has already exited (correctly). I need to make it so that it closes the client's connection when it gets an EOF from the server... Working on that now.

ncdc commented:
Ok, after hacking on this for most of the day, here's where I am...

What I'm struggling with is how to ensure that we're both able to 1) clean up the copy goroutines and release resources appropriately, and 2) make sure all the data flowing between the client and the forwarded port is fully sent and received before cleaning up everything.

Let's examine the flow. When you run `kubectl port-forward -p <pod> 2181`, that creates a new `net.Listener` listening on port 2181 on your local machine (let's call it a laptop). At this point, no connections are established between the laptop and Kubernetes - all it does is set the stage so port forwarding can happen.

When you make a connection to the local listener (telnet/nc/curl/etc), it's at this time that the `kubectl port-forward` process sets up the spdystream connection to the kubelet (via the master). Once that's established, 2 streams are created: 1 for the kubelet to be able to report any errors, and 1 for the actual port forward data. The port-forward process starts 1 goroutine to copy data from the laptop to the pod's port, and 1 to copy data from the pod's port to the laptop. Here's where it gets tricky...

Ideally, both copy routines should finish before we consider the forwarded connection complete. If the laptop->pod copy finishes, we have to wait for the pod->laptop copy to finish before we can reset & close the spdystream stream; if you call stream.Reset() and there is outstanding unread data, you won't be able to read it and it's lost.

The problem is that the laptop->pod copy can finish either because there's no more data to copy (echo foo | nc) or because the connection was closed. If there's no more data to copy, we need to wait for the server to send back anything it has. If the connection was closed, we need to find a way to tell the server to shut down socat, otherwise we'll be leaking processes and goroutines...

smarterclayton commented:
I saw you discussing this this morning - get any further?

On Wed, Aug 5, 2015 at 5:24 PM, Andy Goldstein <notifications@github.com>
wrote:

> Ok, after hacking on this for most of the day, here's where I am...
>
> What I'm struggling with is how to ensure that we're both able to 1) clean
> up the copy goroutines and release resources appropriately, and 2) make
> sure all the data flowing between the client and the forwarded port is
> fully sent and received before cleaning up everything.
>
> Let's examine the flow. When you run kubectl port-forward -p <pod> 2181,
> that creates a new net.Listener listening on port 2181 on your local
> machine (let's call it a laptop). At this point, no connections are
> established between the laptop and Kubernetes - all it does is set the
> stage so port forwarding can happen.
>
> When you make a connection to the local listener (telnet/nc/curl/etc),
> it's at this time that the kubectl port-forward process sets up the
> spdystream connection to the kubelet (via the master). Once that's
> established, 2 streams are created: 1 for the kubelet to be able to report
> any errors, and 1 for the actual port forward data. The port-forward
> process starts 1 goroutine to copy data from the laptop to the pod's port,
> and 1 to copy data from the pod's port to the laptop. Here's where it gets
> tricky...
>
> Ideally, both copy routines should finish before we consider the forwarded
> connection complete. If the laptop->pod copy finishes, we have to wait for
> the pod->laptop copy to finish before we can reset & close the spdystream
> stream; if you call stream.Reset() and there is outstanding unread data,
> you won't be able to read it and it's lost.
>
> The problem is that the laptop->pod copy can finish either because there's
> no more data to copy (echo foo | nc) or because the connection was closed.
> If there's no more data to copy, we need to wait for the server to send
> back anything it has. If the connection was closed, we need to find a way
> to tell the server to shut down socat, otherwise we'll be leaking processes
> and goroutines...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/12283#issuecomment-128153129>
> .
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

ncdc commented:
I've been multitasking and am just trying to get back to this now. I'm still trying to figure out a fool-proof way to handle the following scenarios:

1. `echo foo | nc localhost <port>` - need to make sure that `kubectl port-forward` waits for the full response from the pod's container before closing the connection from `nc`
1. `telnet localhost <port>` but don't send any data - need to make sure that `kubectl port-forward` waits for the full response from the pod's container before closing the connection from `telnet`, taking into account the fact that no data was copied from telnet to the container
1. `telnet localhost <port>', don't send any data, disconnect the telnet session - need dot make sure that `kubectl port-forward` can detect the connection loss and shut down the port-forwarding bits in the kubelet (i.e. socat).

ncdc commented:
I think I may have something that will work. I need to do a bit more testing and then (if all goes well) write some e2es. I'll follow up when I have more for review.

smarterclayton commented:
Maybe we need to get a connection expert in the house.  I get headaches
everytime I try to reason about these sorts of issues :)

On Thu, Aug 13, 2015 at 1:32 PM, Vish Kannan <notifications@github.com>
wrote:

> Assigned #12283 <https://github.com/kubernetes/kubernetes/pull/12283> to
> @smarterclayton <https://github.com/smarterclayton>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12283#event-381716463>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift

ncdc commented:
Updated with more fixes and more tests.

cc @liggitt

ncdc commented:
FYI this PR adds a new image for testing that will need to be pushed to gcr.io (and the PR updated to reflect that).

ncdc commented:
@k8s-bot test this please

liggitt commented on pkg/client/unversioned/portforward/portforward.go:
```diff
@@ -255,16 +256,20 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 	errorChan := make(chan error)
 	doneChan := make(chan struct{}, 2)
 
+	connectionID := util.NewUUID()
```
I'd rather just lock and increment a counter

ncdc replied:
Can do

liggitt commented on pkg/client/unversioned/portforward/portforward.go:
```diff
@@ -279,34 +284,42 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
 		}
+
+		// inform the select below that the remote copy is done
 		doneChan <- struct{}{}
```
this seems asymmetric... we break regardless of error on remote->local copy, but only in the error case on local->remote copy... is there a reason we don't use a wait group and wait for both sides to complete (successfully or with an error)?

ncdc replied:
This was intentional, at least when I was originally working on the fix. However, this code was written prior to the exec fixes that I did with the WaitGroup. I can revisit to see if a WaitGroup would work here too. What is critical is that even if the local->remote copy finishes, the client must wait for the remote->local copy to finish too.

liggitt replied:
yup, I just wasn't sure if we could count on both sides completing (with or without errors), or if this was like the stdin part of remotecommand where we have to fire and forget to avoid hangs

ncdc replied:
To cover what we discussed on irc: with a change to wait for both copies to complete before returning, when running `ab -n 5 -c 1000 ...` where `n` is even a number as small as 5, the local -> remote copies seem to "hang". They don't unblock, and `ab` times out. If I revert to what is in this PR, I can get `n` up to 200 or a bit more. Any larger, and there are issues using the Docker socket to inspect the container (which happens for each locally forwarded connection).

In summary, I think we need to short-circuit and return if there is an error copying local -> remote, but otherwise wait for the remote -> local copy to finish before returning.

pmorie commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+			defer stream.Close()
+
+			streamType := stream.Headers().Get(api.StreamType)
+			glog.V(5).Infof("(%s) handlePortForwardStreams got stream type %s", connectionID, streamType)
```
Nit: I think this message could be a little cleared wrt `connectionID`.  How about:

`"handlePortForwardStreams get stream type %s for connectionID %s"`

ncdc replied:
How would you feel about `(conn %d) ...`? I was thinking it'd be nice to have the conn ID in the same place for the log messages.

Although, if you have 2 different terminals both running `kubectl port-forward` for the same pod port, and you have 2 concurrent connections (1 to each kubectl), then in the kubelet's logs, you could see multiple log lines that all talk about connection ID 0, for example, where some of the lines are for kubectl 1, and others are for kubectl 2, and there's no way to tell which is which. I'm not sure how problematic this may be. If we want to fix it, we need a way to uniquely identify each streaming connection. The debug output in spdystream currently just points the memory address of the connection, which could work for us...

liggitt replied:
yeah, whatever we need to include to identify it uniquely would be helpful. your future self will thank you

pmorie commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
 	port, _ := strconv.ParseUint(portString, 10, 16)
-	err := host.PortForward(pod, uid, uint16(port), dataStream)
+
+	glog.V(5).Infof("(%s) invoking portForwarder.PortForward for port %s", connectionID, portString)
```
Same nit about message context here.

liggitt commented on pkg/client/unversioned/portforward/portforward.go:
```diff
@@ -255,16 +256,20 @@ func (pf *PortForwarder) handleConnection(conn net.Conn, port ForwardedPort) {
+	// we're not writing to this stream
+	errorStream.Close()
 	go func() {
 		message, err := ioutil.ReadAll(errorStream)
```
similar question to remoteCommand... should we always send to the errorChan (nil or an error) and unconditionally receive from errorChan before returning?

ncdc replied:
Yes, I will do that.

liggitt commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -1168,16 +1168,39 @@ func (dm *DockerManager) PortForward(pod *kubecontainer.Pod, port uint16, stream
 	args := []string{"-t", fmt.Sprintf("%d", containerPid), "-n", "socat", "-", fmt.Sprintf("TCP4:localhost:%d", port)}
-	// TODO use exec.LookPath
+
+	_, lookupErr = exec.LookPath("nsenter")
```
is this just to get a better error, or should we be using the looked-up path somehow?

ncdc replied:
I must have had that in a previous revision of this, and somehow it got lost in various edits. I'll put it back in.

liggitt commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -1217,15 +1217,32 @@ func (r *runtime) PortForward(pod *kubecontainer.Pod, port uint16, stream io.Rea
+	if err != nil {
+		return fmt.Errorf("unable to do port forwarding: error creating stdin pipe: %v", err)
+	}
+	go io.Copy(inPipe, stream)
```
should you close inPipe when the copy completes like you do above for docker?

ncdc replied:
Yes, will fix.

liggitt commented on pkg/kubelet/dockertools/manager.go:
```diff
@@ -1168,16 +1168,39 @@ func (dm *DockerManager) PortForward(pod *kubecontainer.Pod, port uint16, stream
+	//
+	// The work around is to use StdinPipe(), as Wait() (called by Run()) closes the pipe
+	// when the command (socat) exits.
+	inPipe, err := command.StdinPipe()
```
make this conditional on `stream != nil`? Was it actually valid to pass a nil stream before, or were the test cases just doing that for expedience?

ncdc replied:
It's not valid to have a nil stdin for port forwarding.

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+				glog.V(5).Infof("(%s) connection ID not in streamChans", connectionID)
+
+				glog.V(5).Infof("(%s) creating stream channel", connectionID)
+				ch = make(chan httpstream.Stream)
```
do we want this to be a buffered channel of size 2 to avoid blocking the main handler? not sure I feel strongly, but wondered

ncdc replied:
It can't hurt, can it?

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+			glog.V(5).Infof("(%s) sending stream to stream channel", connectionID)
+			ch <- stream
+
+			streamLock.Unlock()
```
you can unlock before sending the stream to the channel, rather than after

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+	glog.V(5).Infof("Setting port forwarding streaming connection idle timeout to %v", idleTimeout)
+	conn.SetIdleTimeout(idleTimeout)
 
+	var streamLock sync.Mutex
```
`streamLock := &sync.Mutex{}`, or the pass-by-value to `handlePortForwardStreams` won't actually be using the same lock

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+				streamChans[connectionID] = ch
+
+				glog.V(5).Infof("(%s) starting handlePortForwardStreams", connectionID)
+				go handlePortForwardStreams(podName, uid, ch, portForwarder, connectionID, streamChans, streamLock)
```
I'd prefer the timeout and cleanup handling be done in this function, rather than passing the map, lock and connectionID into `handlePortForwardStreams`... trying to think of cleanest way to do that

liggitt replied:
I guess the timeout handling could stay in `handlePortForwardStreams`, and we can just close and clean up the channel here once `handlePortForwardStreams` returns, e.g.
```
go func(){
  handlePortForwardStreams(podName, uid, ch, portForwarder)
  streamChansLock.Lock()
  defer streamChansLock.Unlock()
  delete(streamChans, connectionID)
  close(ch)
}()
```

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+
+			switch streamType {
+			case api.StreamTypeError:
+				errorStream = stream
```
if errorStream/dataStream are already set, double-assigning should probably be an error

liggitt commented:
maybe split out a commit with the portforward test image so that can get built/pushed?

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+		// old client that isn't generating connectionID
+		streamID := stream.Identifier()
+
+		// When there aren't concurrent new forwarded connections, each
```
if there's an error setting up a stream, it could break the (id/4) rule... I would check the stream type and use Identifier() directly for error streams and Identifier()-2 for data streams

liggitt replied:
and link to the current version of portforward.go to explain why we're computing legacy connectionIDs this way:

https://github.com/kubernetes/kubernetes/blob/843134885e7e0b360eb5441e85b1410a8b1a7a0c/pkg/client/unversioned/portforward/portforward.go#L258-L287

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+		// the client will be unable to forward any more ports until
+		// the `kubectl port-forward` command is terminated and restarted.
+		connectionID = strconv.Itoa(int(streamID) / 4)
+		glog.V(5).Infof("Automatically assigning connection ID %d from stream ID %d", connectionID, streamID)
```
include type in the log message

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
-	defer errorStream.Reset()
+// handlePortForwardStreams waits to receive the error and data streams via streamChan,
+// then uses portForwarder to perform the actual port forwarding logic. If the streams
+// are not received within streamCreationTimeout, this function returns.
```
"this function returns when the port forwarding operation completes, or if the streams are not received from streamChan within StreamCreationTimeout"

pmorie commented:
@liggitt What I've done in the past with test images is have them pushed once the PR merges.  If we need the image pushed for E2E to pass on this, then yeah we should split another commit.

ncdc commented:
@k8s-bot test this please

liggitt commented on pkg/api/types.go:
```diff
@@ -1905,7 +1905,8 @@ const (
 	StreamTypeError  = "error"
 
-	PortHeader = "port"
+	PortHeader         = "port"
```
godoc what these headers are used for

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -587,80 +595,179 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
-		errorStream.Write([]byte(msg.Error()))
+		msg := fmt.Errorf("error forwarding port %d to pod %s, uid %v: %v", port, pod, uid, err)
+		util.HandleError(msg)
+		fmt.Fprintln(errorStream, msg.Error())
```
this adds a newline... was that intentional?

ncdc replied:
No, habit. Fixed in next commit.

Commit:
Port forwarding fixes

Correct port-forward data copying logic so that the server closes its
half of the data stream when socat exits, and the client closes its half
of the data stream when it finishes writing.

Modify the client to wait for both copies (client->server,
server->client) to finish before it unblocks.

Fix race condition in the Kubelet's handling of incoming port forward
streams. Have the client generate a connectionID header to be used to
associate the error and data streams for a single connection, instead of
assuming that streams n and n+1 go together. Attempt to generate a
pseudo connectionID in the server in the event the connectionID header
isn't present (older clients); this is a best-effort approach that only
really works with 1 connection at a time, whereas multiple concurrent
connections will only work reliably with a newer client that is
generating connectionID.

Commit:
Add port-forward-tester image

ncdc commented:
@liggitt @pmorie new port-forward-tester image split into separate commit, everything else squashed down to 1 other commit.

ncdc commented:
@lavalamp I've got a new image in this PR for use with the new port forwarding e2e tests. Are you the right person to coordinate with to get it pushed to gcr.io? Also, if you have time, PTAL at the changes here. 1 thing I think you'll like is I made it so the unit tests and the real code both share the same http handler code for port forwarding. I plan to make the same changes for exec's unit tests as well.

liggitt commented:
I've done a couple review passes, and changes LGTM, but would like one more set of eyes on it

lavalamp commented:
@ncdc I'm super busy-- I'll let @krousey give this a final look. He can build & push the image to gcr.io as needed.

smarterclayton commented on pkg/client/unversioned/portforward/portforward.go:
```diff
@@ -250,67 +254,79 @@ func (pf *PortForwarder) waitForConnection(listener net.Listener, port Forwarded
-	glog.Infof("Handling connection for %d", port.Local)
+	pf.connectionIDLock.Lock()
+	connectionID := pf.nextConnectionID
+	pf.nextConnectionID++
```
This should be abstracted into a method and use defer

ncdc replied:
Will do

ncdc replied:
Actually, is it worth it? Won't I have to hold the lock here, to get the connection ID, and then again in the helper method, to increment it?

liggitt replied:
Make issuing the next id self-contained in a function that locks, defers an unlock, saves the current value, increments, and returns

krousey commented on pkg/client/unversioned/portforward/portforward.go:
```diff
@@ -250,67 +254,79 @@ func (pf *PortForwarder) waitForConnection(listener net.Listener, port Forwarded
+		case len(message) > 0:
+			errorChan <- fmt.Errorf("an error occurred forwarding %d -> %d: %v", port.Local, port.Remote, string(message))
+		default:
+			errorChan <- nil
```
Why write a nil value to the channel when you're just going to close it? Reading from a closed channel returns the default value (nil here).

ncdc replied:
Good point; I'll fix it.

krousey commented on pkg/client/unversioned/portforward/portforward.go:
```diff
@@ -250,67 +254,79 @@ func (pf *PortForwarder) waitForConnection(listener net.Listener, port Forwarded
-		glog.Error(err)
-	case <-doneChan:
+	// wait for either a local->remote error or for copying from remote->local to finish
+	<-doneChan
```
It's a little more verbose, but I think the semantics of what you want might come through better if you use separate channels and select on them. Like

```go
localError := make(chan struct{})
remoteDone := make(chan struct{})
...
go func() {
	...
	close(remoteDone)
}()
...
go func() {
	if _, err := ...; err != nil {
		close(localError)
	}
}()

select {
case <-remoteDone:
case <-localError:
} 

```

liggitt replied:
+1 the double send to a single receive channel bothered me, but nothing cleaner jumped to mind. I like this approach

krousey commented on pkg/kubelet/server.go:
```diff
@@ -620,80 +628,179 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+		case api.StreamTypeData:
+			connectionID = strconv.Itoa(int(stream.Identifier()) - 2)
+		default:
+			connectionID = ""
```
Isn't connectionID already `""` here? Also, is this case worth logging or handling as an error?

krousey commented on pkg/kubelet/server.go:
```diff
@@ -620,80 +628,179 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+			streamLock.Unlock()
+
+			glog.V(5).Infof("(%s) sending stream to stream channel", connectionID)
+			ch <- stream
```
I'm still reading through, but it's not immediately obvious to me that this is guaranteed to happen before the close on this channel on line 734. If this writes to a closed channel, it will panic.

liggitt replied:
so move inside the lock?

krousey replied:
So I think I see several problems with this with misbehaving or malicious clients that could DDoS or outright crash the server.

1. A client could create 3 error streams. The first one is fine. The second one comes in and is written to the buffered channel here. The third one comes in and is blocked on the buffered write. Now the second one is received in `handlePortForwardStreams`, causes an error and an early return, then the channel is closed. Now the write of the third stream to the channel panics.

2. A client creates a data and error stream like normal. `handlePortForwardStreams` is no longer consuming from this channel, so future streams created just build up and leave a lot of goroutines blocked to writing on this channel. These built up goroutines will probably all blow up when `handlePortForwardStreams` returns and closes the channel.

krousey replied:
Yea, moving it inside the lock should be sufficient. Can this connection be reused for port forwarding? Can the channel be recreated and another pair started?

liggitt replied:
Yes, a misbehaving client could create multiple streams with the same connection id... if more than two additional streams came in while the previous streams were still being handled, this would block, but only for this port forward operation, not server-wide. Trying to think if there's a cleaner way to match up error/data streams by connection ID and hand them off to the `handlePortForwardStreams` func

ncdc replied:
This isn't pretty, but we could do something like this. When a new stream is received with a connectionID not currently in `streamChans`, allocate 2 channels: 1 like we're currently doing for informing `handlePortForwardStreams` of new streams for the given connection ID, and another to act as a means for `handlePortForwardStreams` to inform `ServePortForward` (the method that spawns `handlePortForwardStreams`) that it's time to remove the entry from `streamChans`. This would all be within a single hold of the lock, so you wouldn't see any races because of timing.

```go
ch <- stream
bothStreamsReceived := <-statusCh
if bothStreamsReceived {
  delete(streamChans, connectionID)
  // ...
}
```

And in `handlePortForwardStreams`, it would send to `statusCh` after it received each new stream and determined if it's ok to remove the entry from the map.

ncdc replied:
I talked to @liggitt and he suggested trying to come up with something simpler. I'm working on an idea and will post that soon.

liggitt replied:
Sketch at http://play.golang.org/p/THJy5szVRm

Basic idea is to separate the pairing/timeout and the actual port forwarding operation.

Main loop would look more like this:
```
Loop:
	for {
		select {
		case <-conn.CloseChan():
			glog.V(5).Infof("Upgraded connection closed")
			break Loop
		case stream := <-streamChan:
			connectionID := getPortForwardStreamConnectionID(stream)
			glog.V(5).Infof("(%s) Received new stream", connectionID)
			p := getPair(connectionID)
			if complete, err := p.Add(s); err != nil {
				fmt.Println("Error adding stream: ", err)
				removePair(id, p)
			} else if complete {
				fmt.Printf("Completed stream, starting port forwarding for %d\n", id)
				removePair(id, p)
				go handlePortForwardStreams(podName, uid, p.dataStream, p.errorStream, portForwarder, connectionID)
			}
		}
	}
}
```

Timeout is handled when the pair for a given connection id is initialized

krousey replied:
That looks quite a bit better. To more specifically ask my last question, is this use case valid: http://play.golang.org/p/hiBWPV1nuW ?

liggitt replied:
no, that indicates a misbehaving client, so it's not something we have to make work well on the server. with this new approach, we would either pair up data/error streams (if they came in in the right order), or return an error (if two data or two error streams came in for the same connection id).

krousey replied:
Then it would be nice to propagate the completed signal back to the closure `portForwardStreamReeceived` so that it can properly error on this case.

krousey commented on test/images/port-forward-tester/Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+FROM scratch
+ADD portforwardtester portforwardtester
+ADD portforwardtester.go portforwardtester.go
```
Why add the source?

ncdc replied:
I probably wouldn't otherwise but I recently read this comment: https://github.com/kubernetes/kubernetes/pull/13982/files#r39559798

krousey commented:
I build and pushed `gcr.io/google_containers/portforwardtester:1.0`. Sorry that took so long, but you should be able to make those changes now. Please change the Makefile too.

krousey commented on test/e2e/portforward.go:
```diff
@@ -0,0 +1,236 @@
+				{
+					Name: "portforwardtester",
+					//Image: "gcr.io/google_containers/portforwardtester:1.0",
+					Image: "ncdc/portforwardtester:1.0",
```
You should now be able to change this.

krousey commented on test/images/port-forward-tester/Makefile:
```diff
@@ -0,0 +1,15 @@
+	CGO_ENABLED=0 GOOS=linux go build -a -installsuffix cgo -ldflags '-w' ./portforwardtester.go
+
+image: portforwardtester
+	sudo docker build -t ncdc/portforwardtester:$(TAG) .
```
Please remove the `sudo`

krousey commented on test/images/port-forward-tester/Makefile:
```diff
@@ -0,0 +1,15 @@
+	sudo docker build -t ncdc/portforwardtester:$(TAG) .
+
+push: image
+	sudo docker push ncdc/portforwardtester:$(TAG)
```
Please remove the `sudo` and make this `gcloud docker push gcr.io/google_containers/portforwardtester:$(TAG)`

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -620,81 +633,216 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+	h.pairsLock.Lock()
+	defer h.pairsLock.Unlock()
+
+	delete(h.pairs, id)
```
do we care about checking if the pair in the map for that id matches the one we're expecting to remove? not checking means recycled connection ids (which aren't a valid use) could get stomped

ncdc replied:
Good point. If we put this check in, is there any way we could get into a weird state, or does the check just ensure that we don't prematurely remove a recycled id/pair?

liggitt replied:
every pair creation comes with a timeout that will remove it, the check here would just keep us from stomping a recycled id... like I said, I'm not sure I care

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -620,81 +633,216 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+	delete(h.pairs, id)
+}
+
+func (h *portForwardStreamHandler) sessionID(stream httpstream.Stream) string {
```
"session" connotes something that spans requests in my mind... requestID?

ncdc replied:
Gah.... ok

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -620,81 +633,216 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
-				stream.Reset()
+			glog.V(5).Infof("(conn=%p, session=%s) received new stream of type %s", h.conn, sessionID, streamType)
+			p := h.getPair(sessionID)
+			if complete, err := p.add(stream); err != nil {
```
`err == nil && complete`

ncdc replied:
I'm going to add in error handling - just haven't done it yet.

liggitt replied:
k

liggitt commented on pkg/kubelet/server.go:
```diff
@@ -620,81 +633,216 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+			return false, errors.New("data stream already assigned")
+		}
+		p.dataStream = stream
 	}
```
error on unknown types

ncdc replied:
Not needed. See `portForwardStreamReceived`

krousey commented on test/e2e/portforward.go:
```diff
@@ -0,0 +1,235 @@
+)
+
+const (
+	portForwardTestImage = "gcr.io/google_containers/pftest"
```
What is this?

ncdc replied:
Oops :smile: that was the original name for my new image. I'll fix it.

ncdc replied:
Actually, I'll just delete it.

krousey commented on pkg/client/unversioned/portforward/portforward_test.go:
```diff
@@ -295,55 +192,121 @@ func TestGetListener(t *testing.T) {
-		Send     map[uint16]string
-		Receive  map[uint16]string
-		Err      bool
+	//flag.Lookup("v").Value.Set("9")
```
Remove this line please.

ncdc replied:
It will be gone before this PR is ready for merging :-)

On Fri, Sep 25, 2015 at 1:06 PM, krousey <notifications@github.com> wrote:

> In pkg/client/unversioned/portforward/portforward_test.go
> <https://github.com/kubernetes/kubernetes/pull/12283#discussion_r40451966>
> :
>
> >  func TestForwardPorts(t *testing.T) {
> > -	testCases := []struct {
> > -		Upgrader *fakeUpgrader
> > -		Ports    []string
> > -		Send     map[uint16]string
> > -		Receive  map[uint16]string
> > -		Err      bool
> > +	//flag.Lookup("v").Value.Set("9")
>
> Remove this line please.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12283/files#r40451966>.
>

ncdc commented:
@krousey @liggitt updated with a bit of cleanup, some more godoc, and a few more unit tests. Let me know what you think!

ncdc commented:
also @smarterclayton if you have some spare time and want to review

ncdc commented:
@k8s-bot test this

ncdc commented:
@k8s-bot test this

ncdc commented:
@k8s-bot test this

ncdc commented:
@krousey any other comments?

ncdc commented:
Shippable keeps failing with go1.3 with this

```
FAIL	k8s.io/kubernetes/pkg/kubelet	4.402s
[snip]
!!! Error in ./hack/test-go.sh:217
  'return ${rc}' exited with status 1
Call stack:
  1: ./hack/test-go.sh:217 main(...)
Exiting with status 1
```

I don't see any more detail than that. It passes for go1.4.

krousey commented:
Not that it helps, but

```
    FAIL	k8s.io/kubernetes/pkg/kubelet	4.402s
```

I'll rerun it for you.

ncdc commented:
@krousey I've rerun it a few times now already... :cry:

krousey commented:
5 more can't hurt.

ncdc commented:
Ok, I ran with go1.3 locally and can reproduce some data races. I'll try to fix them tonight after dinner.

ncdc commented:
@krousey @liggitt I think I've fixed the data races. PTAL!

krousey commented on pkg/kubelet/server_test.go:
```diff
@@ -1426,3 +1427,355 @@ func TestServePortForward(t *testing.T) {
+		streamCreationTimeout: 100 * time.Millisecond,
+	}
+	p = h.getStreamPair("2")
+	time.Sleep(200 * time.Millisecond)
```
I'm sorry. I should have commented on this way sooner, but can we avoid time.Sleep in tests? This will almost certainly cause a flake at some point or another on Shippable when it's highly contended.

I think the ideal way to do this kind of testing is to change `streamCreationTimeout` from `time.Duration` to a `<-chan time.Time`. In `ServePortForward`, you could initialize it with a `time.After` call and just  select on that in `monitorStreamPair`. Also, I would move the `go h.monitorStreamPair(p)` call to `h.run` after it gets the pair. Then in the tests, you can

```go
h := &portForwardStreamHandler{
	streamPairs:           make(map[string]*portForwardStreamPair),
	streamCreationTimeout: make(chan time.Time),
}
p := h.getStreamPair("1")
handlingDone := make(chan struct{})
go func() {
	h.monitorStreamPair("1")
	close(handlingDone)
}

if !h.hasStreamPair("1") {
	t.Error("This should still be true")
}
close(h.streamCreationTimeout) // explicitly cause a timeout
<-handlingDone // make sure monitorStreamPort ran
if h.hasStreamPair("1") {
	t.Error("should have been removed")
}
```

Also, now the test only takes as long as it needs to. What do you think?

krousey commented on pkg/kubelet/server_test.go:
```diff
@@ -1426,3 +1427,355 @@ func TestServePortForward(t *testing.T) {
+			ok = true
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
```
You can avoid this in the same way as my comment below. Instead of triggering `monitorStreamPair` to complete by causing a timeout, you let it complete by adding both pairs.

ncdc commented:
@krousey updated per your comments. I've removed the stream order test because it relies on sleeps. I have a copy of it if we want to bring it back, but we'd need to figure out a way to do it without sleeping. PTAL

krousey commented:
@ncdc go ahead and squash now. Thank you for your patience with this. :smile:

ncdc commented:
@krousey ok, squashed down to 2 commits... hopefully I did it right (I ended up splitting 1 commit in the middle into 2 so I could put part of it with the commit that adds the portforwardtester image, and the other part of it with the other big commit). I do have a backup copy of the branch pre-squash, if something messed up and we need to recover. PTAL, and thanks for all your time reviewing!

ncdc commented:
@k8s-bot test this

ncdc commented:
@k8s-bot test this

ncdc commented:
After downloading every junit artifact and searching through them for failures, I found this:

```xml
<testsuite tests="8" failures="1" time="5.557" name="k8s.io/kubernetes/pkg/kubelet/network/exec">
		<properties>
			<property name="go.version" value="go1.4.2"></property>
		</properties>
		<testcase classname="exec" name="TestSelectPlugin" time="0.010"></testcase>
		<testcase classname="exec" name="TestSelectVendoredPlugin" time="0.010"></testcase>
		<testcase classname="exec" name="TestSelectWrongPlugin" time="0.000"></testcase>
		<testcase classname="exec" name="TestPluginValidation" time="0.000"></testcase>
		<testcase classname="exec" name="TestPluginSetupHook" time="1.450"></testcase>
		<testcase classname="exec" name="TestPluginTearDownHook" time="0.000">
			<failure message="Failed" type=""></failure>
		</testcase>
		<testcase classname="exec" name="TestPluginStatusHook" time="0.110"></testcase>
		<testcase classname="exec" name="TestPluginStatusHookIPv6" time="0.120"></testcase>
	</testsuite>
```

This doesn't show up in build-log.txt. In fact, that shows this:

```
ok  	k8s.io/kubernetes/pkg/kubelet/network/exec	5.557s
```

@ixdy @quinton-hoole any idea what's going on here with the failure not surfacing?

ncdc commented:
@k8s-bot test this

ncdc commented:
last flake

```
• Failure [321.633 seconds]
Kubectl client
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:682
  Simple pod
  /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:214
    should support exec [It]
    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:188

    Oct  1 07:50:27.236: Error running &{/jenkins-master-data/jobs/kubernetes-pull-build-test-e2e-gce/workspace/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://104.197.17.83 --kubeconfig=/var/lib/jenkins/jobs/kubernetes-pull-build-test-e2e-gce/workspace/.kube/config exec --namespace=e2e-tests-kubectl-mz0x4 nginx echo running in container] []  <nil>  error: pod nginx is not running and cannot execute commands; current phase is Pending
     [] <nil> 0xc20827fde0 exit status 1 <nil> true [0xc20805aff0 0xc20805b018 0xc20805b048] [0xc20805aff0 0xc20805b018 0xc20805b048] [0xc20805b008 0xc20805b040] [0x6b3720 0x6b3720] 0xc2083407e0}:
    Command stdout:
    
    stderr:
    error: pod nginx is not running and cannot execute commands; current phase is Pending
    
    

    /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/util.go:1006
```

ncdc commented:
@krousey squashed & green after several unrelated flakes. PTAL

ixdy commented:
@ncdc regarding https://github.com/kubernetes/kubernetes/pull/12283#issuecomment-144749186, I bet the test failed because of a resource conflict; it uses a plugin named `fmt.Sprintf("test%d", rand.Intn(1000))` in the static path `/tmp/fake/plugins/net`, and because the unit tests on Jenkins are not containerized, it's possible for one test run to affect another.

I'm not sure why
a) there was no useful error message in the JUnit file
b) the test summary indicated a pass when clearly a test failed

Opened #15125 to investigate/track.
